### ISSUE TITLE
google_checks.xml: Public class and interface javadoc is not checked

### ISSUE THREAD

**Issue by fpetry** [checkstyle/checkstyle#5148]
~~~
/tmp$ javac Test.java
~~~

~~~
/tmp$ cat Test.java
public class Test { // No warning about missing Javadoc
}
~~~

~~~
/tmp$ java -Duser.language=en -Duser.country=US -jar checkstyle-8.2-all.jar \
    -c /google_checks.xml Test.java
Starting audit...
Audit done.
~~~

Expected: 
~~~
/tmp$ java -Duser.language=en -Duser.country=US -jar checkstyle-8.2-all.jar \
     -c google_checks.xml Test.java
Starting audit...
[WARN] /tmp/Test.java:1: Missing a Javadoc comment. [JavadocType]
Audit done.
~~~

There should be a warning about missing javadoc for classes and interfaces.

See Google Java Style Guide

> At the minimum, Javadoc is present for every public class, and every public or protected member of such a class, with a few exceptions noted below.

(https://google.github.io/styleguide/javaguide.html#s7.3-javadoc-where-required)

Can be easily achieved by adding

~~~
<module name="JavadocType">
    <property name="scope" value="public"/>
</module>		
~~~

to google_checks.xml

-------

**Migration notes:**

After activation of this Check in config there will be a lot of violation on test classes. Even project like Guava do not write comment on Test classes. We do believe that Google need to update style doc to mention this.
Meanwhile we suggest optional usage of following filter in google config:
```
      <!-- Suppress warnings for test cases from MissingJavadocType, since they may not have javadocs -->
      <module name="SuppressionSingleFilter">
        <property name="files" value=".*test\/|.*Test.*\.java$"/>
        <property name="checks" value="MissingJavadocType"/>
      </module>
```

**Comment by romani** [comment]
@six86 , please read https://github.com/checkstyle/checkstyle/blob/master/.github/ISSUE_TEMPLATE.md and do exactly as we request.

**Comment by six86** [comment]
Updated the description following the template.

**Comment by romani** [comment]
@six86 , thanks a lot for details.
Solution is looks simple, but https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocTypeCheck.java is not a Check that is based on our newly introduced AST for javadoc, so current implementation can cause multiple false positives.
We need to do massive stress testing to make sure we do not case massive amount false positives.

**Comment by romani** [comment]
http://checkstyle.sourceforge.net/google_style.html#a7.3 is marked as fully covered, but it is not correct, image should be changed to blue with reference to this issue in explanation

**Comment by rnveach** [comment]
We added MissingJavadocType in https://github.com/checkstyle/checkstyle/commit/dd45616a0df1ba00e9e62ae9626df5929d94b225 so atleast testing for a javadoc is possible without a huge stress.

**Comment by romani** [comment]
@rnveach , is MissingJavadocType , part of Google style ? Do we cover reported case ? Can we close issue ?

**Comment by rnveach** [comment]
>  is MissingJavadocType , part of Google style ?

JavadocType was never part of Google style, so MissingJavadocType was never added. Only MissingJavadocMethod was added since JavadocMethod was listed there.

Issue seems to be valid to me. All we would need to do is add MissingJavadocType to the config and add some Google UTs to showcase it.

**Comment by romani** [comment]
@fpetry , do you have time to contribute ? We will guide you on what to change.

**Comment by alexbde** [comment]
I may help if you like. I'm interested in realizing Google Style as close as possible.

**Comment by rnveach** [comment]
If no one has started on the issue, anyone is free to start on it once it is approved.

**Comment by wltan** [comment]
I am on it!

**Comment by wltan** [comment]
I believe I am blocked by issue #7747. It so happens that adding `MissingJavadocType` requires support for XPath suppression, which is not supported yet. This test is failing:
```
[ERROR] testViolationsByGoogleAndXpathSuppressions{Capturable, Capturable}  Time elapsed: 0.062 s  <<< FAILURE!
java.lang.AssertionError:
Unexpected output log
Expected: is "Starting audit...\r\nAudit done.\r\n"
     but: was "Starting audit...\r\n[WARN] D:\Workspaces\Git\checkstyle\src\test\resources\com\puppycrawl\tools\checkstyle\main\InputMainViolationsForGoogle.java:3: Missing a Javadoc comment. [MissingJavadocType]\r\nAudit done.\r\n"
        at com.puppycrawl.tools.checkstyle.MainTest.testViolationsByGoogleAndXpathSuppressions(MainTest.java:430)
```
https://github.com/checkstyle/checkstyle/blob/ae9edbd4a8645c8cc1a99ad6890dd0006700d1cb/src/test/java/com/puppycrawl/tools/checkstyle/MainTest.java#L423-L432
https://github.com/checkstyle/checkstyle/blob/3e0e27a4acbda1857f3d89b166075144701e6c98/src/test/resources/com/puppycrawl/tools/checkstyle/main/InputMainViolationsForGoogle.java#L1-L9

Until #7747 is fixed, I don't think there is much I can do right now.

**Comment by wltan** [comment]
There is another problem that I encountered. I tried using the following config:
```xml
        <module name="MissingJavadocType">
            <property name="scope" value="public"/>
            <property name="skipAnnotations" value=""/>
            <property name="tokens" value="CLASS_DEF"/>
        </module>
```
and got this error:
```
[ERROR] testAllCheckTokensAreReferencedInGoogleConfigFile  Time elapsed: 0.005 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: 'MissingJavadocType' should have all acceptable tokens from check in google config or specify an override to ignore the specific tokens ==> expected: <[CLASS_DEF, ENUM_DEF, INTERFACE_DEF, ANNOTATION_DEF]> but was: <[CLASS_DEF]>
        at com.puppycrawl.tools.checkstyle.internal.AllChecksTest.validateAllCheckTokensAreReferencedInConfigFile(AllChecksTest.java:437)
        at com.puppycrawl.tools.checkstyle.internal.AllChecksTest.testAllCheckTokensAreReferencedInGoogleConfigFile(AllChecksTest.java:373)
```
On one hand, it makes sense to restrict tokens to `CLASS_DEF` because that is the exact specification in google style:
> At the minimum, Javadoc is present for every **public class**, and every public or protected member of such a class, with a few exceptions noted below.

But on the other hand, the validation wants us to use all acceptable tokens for some reason.

**Comment by rnveach** [comment]
@wltan This issue is not about xpath support. We just need check added to google config and UTs to show it in action using google config.

> testViolationsByGoogleAndXpathSuppressions

Add javadoc to remove violations that can't be suppressed by xpath.

> wants us to use all acceptable tokens for some reason.

This is not the case. The UT ensures all tokens are used or that they are added to a list of suppressions on why thy don't apply. It ensures that when we add a new token, we review it's use for google as we have been hit by issues before.

See https://github.com/checkstyle/checkstyle/blob/ae9edbd4a8645c8cc1a99ad6890dd0006700d1cb/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java#L177.

There is nothing here that should be blocking this issue that I am seeing.

**Comment by wltan** [comment]
> @wltan This issue is not about xpath support. We just need check added to google config and UTs to show it in action using google config.
> 
> > testViolationsByGoogleAndXpathSuppressions
> 
> Add javadoc to remove violations that can't be suppressed by xpath.
> 
> > wants us to use all acceptable tokens for some reason.
> 
> This is not the case. The UT ensures all tokens are used or that they are added to a list of suppressions on why thy don't apply. It ensures that when we add a new token, we review it's use for google as we have been hit by issues before.
> 
> See
> 
> https://github.com/checkstyle/checkstyle/blob/ae9edbd4a8645c8cc1a99ad6890dd0006700d1cb/src/test/java/com/puppycrawl/tools/checkstyle/internal/AllChecksTest.java#L177
> 
> .
> There is nothing here that should be blocking this issue that I am seeing.

Understood. Thanks a lot for the guidance!

**Comment by pbludov** [comment]
Fix is merged.
